### ISSUE TITLE
Problem with 1D visualization (viz)

### ISSUE THREAD

**Issue by Overcraft90** [pangenome/odgi#394]
Hi,

I'm getting this error while trying to visualize the **hprc_pggb_Y1** graph in `.gfa.gz` format:

> warning [libhandlegraph]: Serialized object does not appear to match deserialzation type.
> warning [libhandlegraph]: It is either an old version or in the wrong format.
> warning [libhandlegraph]: Attempting to load it anyway. Future releases will reject it!
> terminate called after throwing an instance of 'std::length_error'
>   what():  vector::_M_fill_insert
> Aborted (core dumped)

I'm attempting to do so with the _viz_ flag for a 1D visualization; however, I'm not sure whether this is due to a memory limit (I'm working on a mobile workstation) or to the fact I'm not decompressing the graph structure before visualizing. Maybe it is related to something else I'm not aware of. Let me know.

Thanks in advance.

P.S. if necessary I can post the command

**Comment by AndreaGuarracino** [comment]
Hi @Overcraft90,
which is your command line? `odgi` takes as input graphs in `GFA` or `ODGI` format, not compressed files directly.

**Comment by Overcraft90** [comment]
Hi @AndreaGuarracino,

So, my assumption was correct. In particular, I run the following command: `odgi viz -i Downloads/hprc-v1.0-pggb.gfa.gz -o test_pggb -t 16` keeping the graph in a compressed format.

That might have been the issue...

**Comment by AndreaGuarracino** [comment]
Yes, it is. You have three options:

- uncompressed GFA (it will build the ODGI format internally):
```
odgi viz -i Downloads/hprc-v1.0-pggb.gfa -o test_pggb -t 16
```

- uncompressed ODGI file:
```
odgi build -g Downloads/hprc-v1.0-pggb.gfa -o Downloads/hprc-v1.0-pggb.og -t 16
odgi viz -i Downloads/hprc-v1.0-pggb.og -o test_pggb -t 16
```

- compressed ODGI file:
```
odgi build -g Downloads/hprc-v1.0-pggb.gfa -o Downloads/hprc-v1.0-pggb.og -t 16
gzip Downloads/hprc-v1.0-pggb.og
odgi viz -i <(zcat Downloads/hprc-v1.0-pggb.og.gz) -o test_pggb -t 16
```

**Comment by Overcraft90** [comment]
I'm going with the first one (just running the computations). Out of curiosity, I bet it is also the one less demanding, although it might require more storage memory — the graph is huge!!! Almost 100GB itself.

How much is going to take to visualize it with this specs?
![Screenshot from 2022-05-02 15-10-32](https://user-images.githubusercontent.com/98895614/166239192-ad6b0392-824a-4be5-9580-b3144831af3f.png)

**Comment by AndreaGuarracino** [comment]
Unfortunately, `odgi viz` is single-threaded (the internal ODGI format creation is in parallel), but it should not take too much (we are working on a hotter alternative, but it is still in development). The runtime depends on the graph complexity. As a reference, I was able to generate PNGs of 45 whole human haplotypes (all chromosomes) in less than 1 hour.

**Comment by Overcraft90** [comment]
Yes, I tried – I'm not saying it worked; in fact, for some reasons Conda shut abruptly before completion. It might be because of the high memory/processing requirements for the operation.

Possibly, I will have to work on single chromosomes. Is there an easy way to do so?

Thanks.

**Comment by ekg** [comment]
odgi explode will split out the chromosome graphs

**Comment by Overcraft90** [comment]
Thanks again, I will give it a try and see how it works!

**Comment by Overcraft90** [comment]
Hi @ekg & @AndreaGuarracino I got some news,

1. I run `odgi viz`, despite not having indicated the extension for the _-o_ flag, what it returned was an image (I guess) but after opening it I saw what looked like to be a grey rectangle from top to bottom of the image viewer
2. I also run `odgi explode`; however, it quickly saturated the available space I had on the HPC cluster at CINECA (2.5TB), so I've been forced to delete the data so far generated, which were not for the whole pangenome (~7.5h run time and many intermediate files)

Now, I have few questions. Is it correct the output I'm getting from `odgi viz`? Moreover, it would be nice to know roughly how much RAM it takes to run, even though I know it depends on the size of the pangenome. Secondly, as I mentioned, I only have 2.5TB of storage space, how can I run an `odgi explode` without saturating the cluster space I have?

Of course, this also is related to the size of the pangenome, but I was wondering whether it can be somehow controlled with the flag _-b_. I also specified a prefix with _-p_ so that each file was called **chr.X.og**.

Thanks

**Comment by AndreaGuarracino** [comment]
When the `odgi viz` output is particularly large, it is recommended to open the image not with the default visualizer in Ubuntu, but with other tools, such as GIMP (https://www.gimp.org/).

Regarding `odgi explode`, I am surprised. The sum of the exploded components should not be too much larger than the initial graph. Can you try again by adding also `--optimize` to the explode command line? This will optimize the components before writing them to disk and this will also reduce their occupied space. Otherwise, there is something still not clear, or maybe you are hitting a bug.

**Comment by Overcraft90** [comment]
Thanks Andrea,

I will do as you said and let you know!

**Comment by Overcraft90** [comment]
Hi again,

For what concerns `odgi viz` I was able to open the image using Kirita (I think it works in a very similar way to GIMP). However, when it comes to `odgi explode` I think there is something not clear going on, or I might just be hitting a bug. I will report the command line used, maybe you can find some faults:

> odgi explode -i /g100_work/IscrC_PanSV/pangenomes/hprc-v1.0-pggb.gfa -p chr -O -t 48

Also, for you information it might be useful to see a visual output:
![Screenshot from 2022-05-08 19-13-32](https://user-images.githubusercontent.com/98895614/167307624-404492eb-6e41-44a0-8f6c-ebd913ffe42a.png)

**Comment by AndreaGuarracino** [comment]
It depends on the `odgi viz` mode. By default, the colors have no particular meaning, but each path has its own color. Take a look at the following `odgi` tutorial for a bit more information about this: https://odgi.readthedocs.io/en/latest/rst/tutorials/exploratory_analysis.html.

**Comment by Overcraft90** [comment]
Thanks @AndreaGuarracino,

Now, I was wondering is there an easy way to get a 2D visualization of the pangenome? I can see both the `layout` and the `draw` functions do something related to that. Specifically, it seems `layout` generates the _**.lay**_ file to feed to the `draw` function.

I'm not an expert, so are there any suggestions, flags etc. I should use when proceeding down this path? Assuming what I said previously is correct and the pangenome doesn't need any form of pre-processing before generating a 2D visualization.

Thanks

**Comment by ekg** [comment]
If you generate the TSV output from odgi layout, then this can be used
(along with a GFA) as input to gfaestus.

This is a zoomable 2D visualization system. It provides node-level
information, positional information of embedded genomes, and the
possibility of adding BED/GFF3 format annotations.

**Comment by AndreaGuarracino** [comment]
As starting point, you can run the same instructions we use in `pggb` at https://github.com/pangenome/pggb/blob/master/pggb#L572-L584. The instructions I've highlighted will generate the layout (in both `.lay` and `.tsv` formats) and a PNG image of such a layout.

To take a look at the layout interactively with `gfaestus` (https://github.com/chfi/gfaestus), you can run:
```
gfaestus your_graph.gfa your_layout.tsv
```

**Comment by Overcraft90** [comment]
Hi @ekg and @AndreaGuarracino,

Very helpful! I will follow-up with some news as soon as I get some results. Still dubious about the output of `odgi explode` (even after adding the -O flag) though... Anyway, for now I think I will focus on visualization; then, I will possibly explore other functions using a smaller pangenome (five individuals) which I was planning to assemble with `pggb`.

Thanks

**Comment by subwaystation** [comment]
Hi @Overcraft90 Were you able to solve all your problems?

**Comment by Overcraft90** [comment]
No, I'm actually struggling a bit about the first steps to follow in order to assemble a pangenome with `pggb`. Maybe you can have a look at this post [pggb #200](https://github.com/pangenome/pggb/issues/200).

Other than that, I'm now working on a 2D visualization (which is taking a long time to process on the cluster I'm using...), but I'm still facing issues with `odgi explode`, which for now I set aside since I'm focusing on visualization.

I would say, however, my primary need is to assemble a pagnenome, as per [pggb #200](https://github.com/pangenome/pggb/issues/200), because the one I downloaded has been generated by the HPRC and it contains too many haplotypes to make a toy system for me to learn.

Thanks a lot!

**Comment by subwaystation** [comment]
Alright, I think Erik already answered in https://github.com/pangenome/pggb/issues/200.

If you face any problems with the 2D visualization, please open another issue. We recommend to use as many threads as possible. It takes much longer for the 2D layout calculation, because we need more calculations compared to 1D.

If `odgi explode` is still bugging you, could you please open an issue here?

It seems your 1D viz questions are resolved, so I am closing this one.
